skbio/metadata/_interval.py

```diff
@@ -0,0 +1,312 @@
+        >>> from skbio.metadata import IntervalMetadata
+        >>> interval_metadata = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+        ...                       boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+        ...                       boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.query(intervals=[(1, 2)])
+        [Interval(intervals=[(0, 2), (4, 7)], metadata={'name': 'sagA'})]
+
+        """
+        inv_md = Interval(interval_metadata=self,
+                          intervals=intervals,
+                          boundaries=boundaries,
+                          metadata=metadata)
+
+        # Add directly to the tree.  So no need for _is_stale_tree
+        for loc in inv_md.intervals:
+            if loc is not None:
+                start, end = loc
+                self._intervals.add(start, end, inv_md)
+
+        self._metadata.append(inv_md)
+
+    def _rebuild_tree(self, intervals):
+        self._intervals = IntervalTree()
+        for f in intervals:
+            for inv in f.intervals:
+                start, end = inv
+                self._intervals.add(start, end, f)
+
+    def _query_interval(self, interval):
+        start, end = _polish_interval(interval)
+        invs = self._intervals.find(start, end)
+        return invs
+
+    def _query_attribute(self, intervals, metadata):
+        if metadata is None:
+            return []
+
+        queries = []
+        for inv in intervals:
+            for (key, value) in metadata.items():
+                if inv[key] != value:
+                    continue
+                queries.append(inv)
+        return queries
+
+    def query(self, intervals=None, boundaries=None, metadata=None):
+        """ Looks up Interval objects with the intervals, boundaries and keywords.
+
+        Parameters
+        ----------
+        intervals : iterable of tuple of ints
+            A list of intervals associated with the Interval object.
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the Interval object.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Returns
+        -------
+        list, Interval
+            A list of Intervals satisfying the search criteria.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> interval_metadata = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+        ...                       boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+        ...                       boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.query(intervals=[(1, 2)])
+        [Interval(intervals=[(0, 2), (4, 7)], metadata={'name': 'sagA'})]
```

Comment:
Found out that my assert statements were missing for some reason :L

That is now fixed.  Thanks!